Fix bug in weight-based sort

nexbridge opened the pull request:
**Summary**
Quotient used in random number generation in weight_based_sort() function should always be less than 1 to avoid unexpected behaviour

**Type**
Bug fix

**Solution**
- Changed RAND_MAX to RAND_MAX + 1 so that the quotient will always be less than 1
- Changed casting from float to double to avoid rounding errors, particularly when the value of rand() nears RAND_MAX and thus the quotient could round to 1 unintentionally

**Compatibility**
The proposed solution is a strict improvement with no risk of side-effects

bogdan-iancu commented:
Hi @nexbridge , thanks for your contribution. Indeed the missing `+1` may lead to `bug in weight sort` situations if the generated number `rand_no` is the same as `weight_sum`. But this may be fixed by changing the `if (running_sum[i]>rand_no)` condition to `>=` also. IMHO this will be more intuitive (basically shifting a bit the matching intervals for each destination to also include the its own running sum.
On the float versus double, agreed!

nexbridge commented:
Hi @bogdan-iancu,

Ah OK - should I close this then and leave you to apply your own bug fix?

Thanks

bogdan-iancu commented:
@nexbridge if you could update the PR to replace that `+1` with change in the test, it will be super, I will accept it.

nexbridge commented:
@bogdan-iancu OK I've removed the `+1` now and changed the `>` to `>=` as suggested, although I haven't checked your logic on that one so am relying on you! 🙂

bogdan-iancu commented:
perfect, thank you. Once the checks/tests will complete, I will merge it

bogdan-iancu commented:
Merged and backported, thank you @nexbridge
